README.md
# errobj [![](https://img.shields.io/npm/v/errobj.svg)](https://www.npmjs.com/package/errobj) [![](https://img.shields.io/badge/source--000000.svg?logo=github&style=social)](https://github.com/omrilotan/errobj)

## ☠️ Serialise errors to literal (JSONable) object

- ✔︎ Designed for error loggers
- ✔︎ Serialises errors to literal objects
- ✔︎ Supports **any** properties attached to the error
- ✔︎ Expands the error details with lineNumber, columnName, fileName, functionName, ...
- ✔︎ Parses [error cause](https://github.com/tc39/proposal-error-cause)
- ✔︎ Accepts an enrichment object
- ✔︎ Parses the stack trace ([error-stack-parser](https://www.npmjs.com/package/error-stack-parser))
- ✔︎ Isomorphic

### TL;DR

```js
import { errobj } from 'errobj';

try {
	some broken code
} catch (error) {
	send(JSON.stringify(errobj(error)));
}
```

#### Arguments

1. `{Error}` (error) An error to be serialised
2. `{Object}` (enrichment) [_optional_] - This object's field values will be assigned to the serialised error
3. `{Object}` (options) [_optional_, _nullable_] - See details below
   - `{Number}` offset [_optional_] - Offset the parsed stack, and error position details. Good for middleware created error objects.
   - `{Number}` parsedStack [_optional_] - Add a parsed stack of the error with a certain depth

### Example: Sending uncaught error to an HTTP error logger

```js
const { errobj } = require("errobj");

const original_onerror = window.onerror; // play nicely
window.onerror = function (message, url, lineNumber, columnNumber, error) {
	fetch("/error-logger", {
		method: "POST",
		body: JSON.stringify(
			errobj(error, { message, url, lineNumber, columnNumber, level: "error" })
		),
	});

	return original_onerror(message, url, lineNumber, columnNumber, error);
};
```

## Examples

### The serialised error

```js
{
	name: 'RangeError',
	message: 'Nothing',
	stack: 'ReferenceError: something is not defined\nat change (index.html:46)\nat index.html:53\nat index.html:56',
	lineNumber: '46',
	columnNumber: '12',
	fileName: 'index.html',
	functionName: 'change',
	source: 'at change (index.html:46)',
	level: 'error'
}
```

### Add fields to the parsed object

```js
errobj(error, { flow: "registration" });
```

### option: offset (stack)

```js
function verboseLog(message) {
	const error = new Error(message);
	send(errobj(error, null, { offset: 1 }));
}
```

### option: parsedStack

```js
errobj(error, null, {parsedStack: Infinity});

{
	...
	parsedStack: [
		{
			lineNumber: 46,
			fileName: 'index.html',
			functionName: 'change',
			source: 'at change (index.html:46)'
		},
		{
			lineNumber: 53,
			fileName: 'index.html',
			source: 'at index.html:53'
		},
		{
			lineNumber: 56,
			fileName: 'index.html',
			source: 'at index.html:56'
		}
	],
	...
}
```


CHANGELOG.md
# CHANGELOG

## 3.0.1

- Fix: error.cause should remain its original structure instead of being "stringified"

## 3.0.0

- New builds for commonjs and esm
- export is now named

```diff
- import errobj from 'errobj'
+ import { errobj } from 'errobj'
```

## 2.4.5

- Fix: error.cause should remain its original structure instead of being "stringified" (backport from 3.0.1)

## 2.4.4

- Workaround for [ErrorStackParser.parse is not a function](https://github.com/stacktracejs/error-stack-parser/issues/80)

## 2.4.3

- Add `babel` and ensure bundle is `es5` compliance.

## 2.4.2

- Replace parcel build with webpack

## 2.4.1

- [FIX] Don't add cause field when cause is undefined

## 2.4.0

- Support error.cause, add "description" custom field

## 2.3.0

- Add more fields to error

## 2.2.1

- Move repo - no change

## 2.2.0

- Support error.toJSON

## 2.1.0

- Parsed stack specific depth instead of just with or without

## 2.0.0

- Breaking: parsedStack is not attached by default
- Add stack offset

## 1.0.0

- Stabailse package - no change

## 0.0.0

- Serialise errors to literal (JSONable) object


spec.js
const { expect } = require('chai');
const errobj = require('.');

const [ major, minor ] = process.versions.node.split('.').map(Number);
const supportsCause = major > 16 || major === 16 && minor >= 9;

describe('error-notation', () => {
	it('Should convert an error to an object', () => {
		const error = new Error('Everything is okay');

		const obj = errobj(error);
		expect(obj.toString()).to.equal('[object Object]');
	});
	it('Should include all error fields', () => {
		const error = new TypeError('Nothing');
		error.details = { answer: 42 };
		error.code = 'UNKNERR';
		const obj = errobj(error);
		expect(obj.message).to.equal('Nothing');
		expect(obj.name).to.equal('TypeError');
		expect(obj.code).to.equal('UNKNERR');
		expect(obj.details).have.any.keys('answer');
		expect(obj.stack).to.be.a('string');
	});

	it('Should include any custom property attached to the error', () => {
		const error = new RangeError('Nothing');
		error.extra = 'information';
		const { extra } = errobj(error);
		expect(extra).to.equal('information');
	});
	it('Should include enrichment fields', () => {
		const error = new RangeError('Nothing');
		const { extra } = errobj(error, { extra: 'information' });
		expect(extra).to.equal('information');
	});
	it('Should give precedence to enrichment fields over the native ones', () => {
		const error = new RangeError('Nothing');
		const { message } = errobj(error, { message: 'Something' });
		expect(message).to.equal('Something');
	});

	it('Should find line and column from browser error stack', () => {
		const error = new RangeError('Nothing');
		error.stack = `ReferenceError: something is not defined
at change (index.html:46)
at index.html:53
at index.html:56`;
		const { fileName, lineNumber, columnNumber } = errobj(error);
		expect(fileName).to.equal('index.html');
		expect(lineNumber).to.equal(46);
		expect(columnNumber).to.be.undefined;
	});

	it('Should not attach parsedStack by default', () => {
		const error = new RangeError('Nothing');
		error.stack = `ReferenceError: something is not defined
at change (index.html:46)
at index.html:53
at index.html:56`;
		const { parsedStack } = errobj(error);
		expect(parsedStack).to.be.undefined;
	});

	it('Should attach specific lines to parsedStack to the details', () => {
		const error = () => {
			const err = new RangeError('Nothing');
			err.stack = `ReferenceError: something is not defined
at change (index.html:46)
at index.html:53
at index.html:56`;
			return err;
		};
		expect(errobj(error(), null, { parsedStack: true }).parsedStack).to.have.lengthOf(3);
		expect(errobj(error(), null, { parsedStack: Infinity }).parsedStack).to.have.lengthOf(3);
		expect(errobj(error(), null, { parsedStack: 2 }).parsedStack).to.have.lengthOf(2);
		expect(errobj(error(), null, { parsedStack: false }).parsedStack).to.be.undefined;
	});

	it('Should find line and column from nodejs error stack', () => {
		const error = new RangeError('Nothing');
		error.stack = `at /app/dist/apps/listings/server.js:1329:40
at Array.filter (<anonymous>)
at Object.category (/app/dist/apps/listings/server.js:1327:37)
at buildSubCategoryFilter (/app/dist/apps/listings/server.js:40171:31)
at buildAppFilters (/app/dist/apps/listings/server.js:40145:18)
at Object.listingsResult (/app/dist/apps/listings/server.js:29813:22)
at /app/dist/apps/listings/server.js:27177:21
at process._tickCallback (internal/process/next_tick.js:68:7)`;
		const { fileName, lineNumber, columnNumber } = errobj(error);
		expect(fileName).to.equal('/app/dist/apps/listings/server.js');
		expect(lineNumber).to.equal(1329);
		expect(columnNumber).to.equal(40);
	});

	it('Should find line and column from browser error stack', () => {
		const error = new RangeError('Nothing');
		error.stack = `TypeError: Cannot read property 'gf' of undefined
at t.r.getPageLoadTime (https://cdn.website.com/assets/application.js:1:284663)
at d (https://cdn.website.com/assets/business-logic.js:1:286145)
at https://connect.facebook.net/en_US/fbevents.js:25:21849
at HTMLIFrameElement.b (https://connect.facebook.net/en_US/fbevents.js:24:3061)`;
		const { fileName, lineNumber, columnNumber } = errobj(error);
		expect(fileName).to.equal('https://cdn.website.com/assets/application.js');
		expect(lineNumber).to.equal(1);
		expect(columnNumber).to.equal(284663);
	});

	it('Should prefer existing lineNumber and columnNumber', () => {
		const error = new RangeError('Nothing');
		error.lineNumber = 2;
		error.columnNumber = 4;
		error.stack = `TypeError: Cannot read property 'gf' of undefined
at t.r.getPageLoadTime (https://cdn.website.com/assets/application.js:1:284663)
at d (https://cdn.website.com/assets/business-logic.js:1:286145)
at https://connect.facebook.net/en_US/fbevents.js:25:21849
at HTMLIFrameElement.b (https://connect.facebook.net/en_US/fbevents.js:24:3061)`;
		const { lineNumber, columnNumber } = errobj(error);
		expect(lineNumber).to.equal(2);
		expect(columnNumber).to.equal(4);
	});

	it('Should offset the parsed stack trace', () => {
		const error = new RangeError('Nothing');
		error.stack = `TypeError: Cannot read property 'gf' of undefined
at t.r.getPageLoadTime (https://cdn.website.com/assets/application.js:1:284663)
at d (https://cdn.website.com/assets/business-logic.js:4:286145)
at https://connect.facebook.net/en_US/fbevents.js:25:21849
at HTMLIFrameElement.b (https://connect.facebook.net/en_US/fbevents.js:24:3061)`;
		let lineNumber, columnNumber;
		({ lineNumber, columnNumber } = errobj(error, null, { offset: 1 }));
		expect(lineNumber).to.equal(4);
		expect(columnNumber).to.equal(286145);

		({ lineNumber, columnNumber } = errobj(error, null, { offset: 2 }));
		expect(lineNumber).to.equal(25);
		expect(columnNumber).to.equal(21849);
	});

	it('Should support errors with a toJSON function', () => {
		class CustomError extends Error {
			toJSON() {
				return {
					message: 'Custom wrong thing',
					stack: this.stack,
				};
			}
		}

		const error = new CustomError('Something must have gone terribly wrong');
		error.stack = `TypeError: Cannot read property 'gf' of undefined
at t.r.getPageLoadTime (https://cdn.website.com/assets/application.js:1:284663)
at d (https://cdn.website.com/assets/business-logic.js:1:286145)
at https://connect.facebook.net/en_US/fbevents.js:25:21849
at HTMLIFrameElement.b (https://connect.facebook.net/en_US/fbevents.js:24:3061)`;

		const { message, lineNumber, columnNumber, extra } = errobj(error, { extra: 'additional info' });
		expect(message).to.equal('Custom wrong thing');
		expect(lineNumber).to.equal(1);
		expect(columnNumber).to.equal(284663);
		expect(extra).to.equal('additional info');
	});
	supportsCause && it('should print cause string', () => {
		const error = new Error('Something must have gone terribly wrong', { cause: 'something horrible' });
		const { cause } = errobj(error);
		expect(cause).to.equal('something horrible');
	});
	supportsCause && it('should print cause string', () => {
		const error = new Error('Something must have gone terribly wrong');
		expect(Object.keys(errobj(error))).not.to.include('cause');
	});
	supportsCause && it('should parse cause error', () => {
		const err = new Error('something horrible');
		const error = new Error('Something must have gone terribly wrong', { cause: err });
		const original = errobj(err);
		const { cause } = errobj(error);
		expect(cause).to.equal(JSON.stringify(original));
	});
	supportsCause && it('should escape circular reference in cause', () => {
		const error = new Error('Something must have gone terribly wrong');
		error.cause = error;
		const { cause } = errobj(error);
		expect(cause).to.equal('[Circular]');
	});
});
